### PR TITLE
Refactor GameState using new GameStateHistory

### PR DESCRIPTION
**Pull request by audebert**
Changes:

- No more naked pointers to GameState, only `std::unique_ptr`s. This simplifies
  the code and make mistakes harder to make.
- rules::Action do not handle GameState anymore, they only have to know how to
  mutate the GameState they are given.
- Api is now responsible to call `game_state_.save()` before applying each
  action. This makes a copy of the current GameState and store it to be used if
  cancel() is called.
- Rules do not have GameState anymore, because they cannot use cancel().
- GameState object are held by a utils::VersionedPtr container, which knows how
  to save and restore GameState instances.

Performance: overall equivalent (same amount of GameState copies), better when
the caller knows that they do not have to support cancel() (no copies are made
if they don't call `game_state.save()`)

Also:

- Replace typedef with using

### CODE REVIEW

File: tools/generator/files/rules.cc [hunk 1]
```diff
@@ -1,9 +1,13 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 #include "rules.hh"
+
+#include <memory>
+#include <utility>
+
 #include "actions.hh"
 
 Rules::Rules(const rules::Options opt)
-    :!!rules_cc_inheritance!! sandbox_(opt.time)
```
Comment (seirl) [inline tools/generator/files/rules.cc:7]: Is that intended? I imagine this would break the template replacement.

**Review by seirl — CHANGES_REQUESTED** [review]

File: tools/generator/files/rules.cc (hunk 1 above)
Comment (audebert) [reply tools/generator/files/rules.cc:7]: This was not intentional. Thanks.

**Comment by audebert** [comment]
I've been thinking of making a specialized implementation of utils::VersionedPtr named GameStateHistory that allow re-introducing the GameState::copy() method and allow hiding the copy constructor. I'll update the PR with that change.

**Review by seirl — APPROVED** [review]

**Review by Shaac — APPROVED** [review]

**Comment by audebert** [comment]
Now updating the games. Once this is done, will update everything at once.

**Pull request merged**